{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Input genomic fasta file: Scer_SGD.fna\n",
      " Input gff file: Scer_SGD.gff\n",
      "Building structured data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 351379/351379 [00:01<00:00, 269144.45it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding Supplemental Attribute table...\n",
      "Finding unique attribute keys...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 351377/351377 [00:01<00:00, 318773.00it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making attribute table...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 351377/351377 [00:03<00:00, 111395.98it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "date Tue Jan 13 13:06:13 2015\n",
      " Created by Saccharomyces Genome Database (http://www.yeastgenome.org/)\n",
      " Weekly updates of this file are available for download from:\n",
      " http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff\n",
      "Building structured data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 23076/23076 [00:00<00:00, 301025.29it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding Supplemental Attribute table...\n",
      "Finding unique attribute keys...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 23058/23058 [00:00<00:00, 194367.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making attribute table...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 23058/23058 [00:00<00:00, 88183.10it/s]\n"
     ]
    }
   ],
   "source": [
    "import gff3_parser\n",
    "import polars as pl\n",
    "\n",
    "orfs = pl.from_pandas(gff3_parser.parse_gff3(\"input/mapping_orf_Scer_SGD_noMT.gff\", parse_attributes=True))\n",
    "\n",
    "GFF = gff3_parser.parse_gff3(\"input/Scer.gff\", parse_attributes=True)\n",
    "\n",
    "\n",
    "COLUMNS = [[\"seq_id\", \"start\", \"end\", \"strand\", \"phase\", \"attributes\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import OrderedDict as OD\n",
    "import polars as pl\n",
    "import warnings\n",
    "import re\n",
    "\n",
    "GFF_POLARS = pl.from_pandas(GFF)\n",
    "\n",
    "ORF_DF_COLUMNS = ['Seqid',\n",
    " 'Source',\n",
    " 'Type',\n",
    " 'Start',\n",
    " 'End',\n",
    " 'Score',\n",
    " 'Strand',\n",
    " 'Phase',\n",
    " 'Status',\n",
    " 'color',\n",
    " 'Parent',\n",
    " 'ID',\n",
    " 'Ovp_with',\n",
    " 'Ovp_gene']\n",
    "\n",
    "class Gene:\n",
    "\n",
    "    def __init__(self, ID, chromosome, start, end, multi, sense):\n",
    "\n",
    "        self.ID = ID\n",
    "        self.chromosome = chromosome\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.sense = sense\n",
    "        self.multi = multi\n",
    "        self.aORFs = OD()\n",
    "        self._exons = OD()\n",
    "\n",
    "    @property\n",
    "    def exons(self):\n",
    "        return self._exons\n",
    "\n",
    "    @exons.setter\n",
    "    def exons(self, value):\n",
    "    \n",
    "        self._exons = value\n",
    "\n",
    "    def add_exon(self, key, value):\n",
    "        \n",
    "        if key in self._exons:\n",
    "            raise KeyError(f'exon {key} already exists')\n",
    "        else:\n",
    "            self._exons[key] = value\n",
    "\n",
    "    def sort(self, feature):\n",
    "\n",
    "        if isinstance(feature, list):\n",
    "            \n",
    "\n",
    "    \n",
    " \n",
    "class Orf: \n",
    "\n",
    "    def __init__(self, ID, start, end, gene ):\n",
    "\n",
    "        self.ID = ID\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.gene = gene\n",
    "        self._frame = None\n",
    "        self._ribospike = None\n",
    "\n",
    "    @property\n",
    "    def ribospike(self):\n",
    "        return self._ribospike\n",
    "    \n",
    "    @ribospike.setter\n",
    "    def ribospike(self, value):\n",
    "        if type(value) != int:\n",
    "            raise TypeError('ribospike must be int')\n",
    "        else:\n",
    "            self._ribospike = value\n",
    "            self._frame = self.gene.start + self._ribospike\n",
    "\n",
    "    @property\n",
    "    def frame(self):\n",
    "        return self._frame\n",
    "\n",
    "\n",
    "class GeneStructureError(Exception):\n",
    "    def __init__(self, message):\n",
    "        self.message = message\n",
    "        super().__init__(self.message)\n",
    "\n",
    "\n",
    "def get_good_column_names(columns):\n",
    "\n",
    "    import re\n",
    "    from itertools import chain\n",
    "\n",
    "    parent_col = [col for col in columns if re.match(r\"[Pp]arent\", col)]\n",
    "    name_col = [col for col in columns if re.match(r\"[Nn]ame\", col)]\n",
    "\n",
    "    if len(parent_col) == 1 and len(name_col) == 1:\n",
    "\n",
    "        return str(parent_col[0]), str(name_col[0])\n",
    "    \n",
    "    else:\n",
    "\n",
    "        raise AttributeError(\"Problem with GFF columns : Parent or Name not found. See get_good_column_names()\")\n",
    "        \n",
    "def return_gene_infos(gene_infos, all_exons = True) -> dict:\n",
    "\n",
    "\n",
    "\n",
    "    chromosome = gene_infos[\"Seqid\"].unique().to_list()[0],\n",
    "    start = gene_infos.filter(pl.col(\"Type\") == \"gene\")[\"Start\"].to_list()[0],\n",
    "    end = gene_infos.filter(pl.col(\"Type\") == \"gene\")[\"End\"].to_list()[0],\n",
    "    sense = gene_infos.filter(pl.col(\"Type\") == \"gene\")[\"Strand\"].to_list()[0]\n",
    "    multi = gene_infos[\"Type\"].to_list().count(\"CDS\") > 1\n",
    "\n",
    "\n",
    "    return {\n",
    "\n",
    "        \"chromosome\" : chromosome,\n",
    "        \"start\" : start,\n",
    "        \"end\" : end,\n",
    "        \"sense\" : sense,\n",
    "        \"multi\" : multi\n",
    "    } \n",
    "\n",
    "def init_gene_object(gene_id, gff_dataframe):\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    On veut pour la gene_id donnée initiliaser un objet Gene avec ses attributs\n",
    "    tirés du gff_dataframe.\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    pattern = fr\".*{gene_id}.*\"\n",
    "\n",
    "    parent, name = get_good_column_names(gff_dataframe.columns)\n",
    "\n",
    "    # Polars does not support regex filtering : pandas is used instead\n",
    "    gene_rows = pl.from_pandas(gff_dataframe[\n",
    "        gff_dataframe[parent].str.contains(pattern, regex=True, na=False) |\n",
    "        gff_dataframe[name].str.contains(pattern, regex=True, na=False)\n",
    "    ]) # Get all rows related to the gene being initialized\n",
    "\n",
    "\n",
    "    \n",
    "    gene_infos = return_gene_infos(gene_rows)\n",
    "\n",
    "\n",
    "    gene = Gene(\n",
    "\n",
    "        ID = gene_id,\n",
    "        chromosome = gene_infos[\"chromosome\"],\n",
    "        start = gene_infos[\"start\"],\n",
    "        end = gene_infos[\"end\"],\n",
    "        sense = gene_infos[\"sense\"],\n",
    "        multi = gene_infos[\"multi\"]\n",
    "        \n",
    "    )\n",
    "    \n",
    "\n",
    "    cds_counter = 1\n",
    "    for exon in gene_rows.filter(pl.col(\"Type\") == \"CDS\").iter_rows(named = True):\n",
    "\n",
    "        key = f'{exon[\"ID\"]}-{cds_counter}' if exon[\"ID\"] else f'{exon[\"Name\"]}-{cds_counter}'\n",
    "        \n",
    "        gene.add_exon(\n",
    "            key = key,\n",
    "            value = OD(\n",
    "                [\n",
    "                    (\"Start\", exon[\"Start\"]),\n",
    "                    (\"End\", exon[\"End\"]),\n",
    "                    \n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "\n",
    "        cds_counter = cds_counter + 1\n",
    "\n",
    "\n",
    "\n",
    "    return gene\n",
    "\n",
    "\n",
    "\n",
    "def check_double_overlap(row : tuple):\n",
    "\n",
    "    orf = dict(zip(ORF_DF_COLUMNS, row))\n",
    "    \n",
    "    overlaps = [match for item in orf[\"Ovp_with\"].split(\"|\") for match in re.findall(r\"\\b([\\w-]+)_mRNA\\b\", item)]\n",
    "\n",
    "    if len(overlaps) == 0:\n",
    "\n",
    "        overlaps = [match for item in orf[\"Ovp_with\"].split(\"|\") for match in re.findall(r\"\\b([\\w-]+)_CDS\\b\", item)]\n",
    "    \n",
    "    buffer = []\n",
    "\n",
    "    if len(overlaps) != 1:\n",
    "\n",
    "        for overlap in overlaps:\n",
    "\n",
    "            \n",
    "            if GFF_POLARS.filter(\n",
    "                \n",
    "                (pl.col(\"ID\") == overlap)\n",
    "                )[\"Strand\"].unique().to_list()[0] == orf[\"Strand\"]:\n",
    "\n",
    "                buffer.append(overlap)\n",
    "                \n",
    "\n",
    "        if len(buffer) == 1: # Several genes are overlapped by the ORF, but only one is on the same strand\n",
    "\n",
    "            orf[\"Ovp_gene\"] = buffer[0]\n",
    "            return tuple(orf.values())\n",
    "        \n",
    "    \n",
    "        elif len(buffer) == 0: # No gene found on the same strand as the ORF\n",
    "\n",
    "            orf[\"Ovp_gene\"] = \"NA\"\n",
    "            return tuple(orf.values())\n",
    "        \n",
    "        else: # Several genes found on the same strand as the ORF\n",
    "\n",
    "            orf[\"Ovp_gene\"] = \"Two_or_more_genes\"\n",
    "            return tuple(orf.values())\n",
    "\n",
    "\n",
    "    else: # If there is only one gene found in the overlapping information given by ORFMine ID\n",
    "\n",
    "        orf[\"Ovp_gene\"] = overlaps[0]\n",
    "        return tuple(orf.values())\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Décider de la structure de de chaque ORF/Exon + de la structure qui les contient \n",
    "Comment implémenter ?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern = r'\\b([\\w-]+)_CDS\\b'\n",
    "\n",
    "dframe = (\n",
    "    orfs\n",
    "    .filter(pl.col(\"Type\") == \"nc_ovp_same-CDS\")\n",
    "    .with_columns([\n",
    "        pl.col(\"ID\").apply(lambda value: value.split(\"_\")[3]).alias(\"Phase\"), # Extract strand phase from the ID generated by ORFMine\n",
    "        pl.lit(\"NA\").alias(\"Ovp_gene\") # Create a column for the gene ID that will be filled by check_double_overlap()\n",
    "    ])\n",
    ")\n",
    "\n",
    "\n",
    "dframe = dframe.apply(check_double_overlap) # Use .apply() method to leverage parallelization\n",
    "dframe.columns = ORF_DF_COLUMNS\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped = dframe.groupby(\"Ovp_gene\")\n",
    "\n",
    "gene_list = list()\n",
    "for name,data in grouped:\n",
    "\n",
    "    if \"gene\" in GFF_POLARS.filter(pl.col(\"ID\") == name)[\"Type\"].unique().to_list(): # If overlapped feature is not a gene ( = transposable for example ) it's not stored\n",
    "\n",
    "        gene = init_gene_object(name, GFF)\n",
    "\n",
    "        for row in data.iter_rows(named = True):\n",
    "\n",
    "            gene.add_orf(\n",
    "                key = row[\"ID\"],\n",
    "                value = OD(\n",
    "                    [\n",
    "                        (\"Start\", row[\"Start\"]),\n",
    "                        (\"End\", row[\"End\"]),\n",
    "                        (\"Strand\", row[\"Strand\"]),\n",
    "                        (\"Phase\", row[\"Phase\"]),\n",
    "                        (\"Ovp_with\", row[\"Ovp_with\"])\n",
    "                    ]\n",
    "                )\n",
    "            )\n",
    "    \n",
    "        "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\"\"\"\n",
    "for name, data in grouped:\n",
    "\n",
    "\n",
    "    print(name)\n",
    " \n",
    "    gene = init_gene_object(name, gff)\n",
    "    \n",
    "    if gene:\n",
    "        print(gene.get_exons())\n",
    "        print(gene.sense)\n",
    "        \n",
    "\n",
    "    # for orf in data.iter_rows(named = True):\n",
    "\n",
    "        # print(orf)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrderedDict([('a', 1), ('b', 2)])\n",
      "OrderedDict([('b', 2), ('a', 1)])\n"
     ]
    }
   ],
   "source": [
    "def return_gene_infos(gene_infos : polars.Dataframe, all_exons = True):\n",
    "\n",
    "\n",
    "    chromosome = gene_infos[\"Seqid\"].unique().to_list()[0],\n",
    "    start = gene_infos.filter(pl.col(\"Type\") == \"gene\")[\"Start\"].to_list()[0],\n",
    "    end = gene_infos.filter(pl.col(\"Type\") == \"gene\")[\"End\"].to_list()[0],\n",
    "    sense = gene_infos.filter(pl.col(\"Type\") == \"gene\")[\"Strand\"].to_list()[0]\n",
    "    multi = len(gene_infos[\"Type\"].to_list().count(\"mRNA\") > 1)\n",
    "\n",
    "def check_gene_structure(gene_infos : polars.DataFrame):\n",
    "\n",
    "    try:\n",
    "        exon_or_CDS = gene_infos[\"Type\"].to_list().count(\"CDS\") == gene_infos[\"Type\"].to_list().count(\"exon\")\n",
    "        num_mRNA = gene_infos[\"Type\"].to_list().count(\"mRNA\")\n",
    "        \n",
    "        if not exon_or_CDS:\n",
    "            raise GeneStructureError(\"The counts of 'CDS' and 'exon' in the 'Type' column do not match.\")\n",
    "        if num_mRNA != 1:\n",
    "            raise GeneStructureError(\"The count of 'mRNA' in the 'Type' column is not 1.\")\n",
    "\n",
    "        # Add more error checks here as needed in the future.\n",
    "\n",
    "        return 0\n",
    "\n",
    "    except GeneStructureError as e:\n",
    "        return e\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[OrderedDict([('Start', 10), ('End', 15)]), OrderedDict([('Start', 20), ('End', 30)]), OrderedDict([('Start', 25), ('End', 35)])]\n"
     ]
    }
   ],
   "source": [
    "from collections import OrderedDict\n",
    "\n",
    "# List of OrderedDict objects\n",
    "data_list = [\n",
    "    OrderedDict([('Start', 20), ('End', 30)]),\n",
    "    OrderedDict([('Start', 10), ('End', 15)]),\n",
    "    OrderedDict([('Start', 25), ('End', 35)]),\n",
    "]\n",
    "\n",
    "# Sort the list based on the 'Start' key\n",
    "sorted_list = sorted(data_list, key=lambda x: x['Start'])\n",
    "\n",
    "# Print the sorted list\n",
    "print(sorted_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = Gene(\n",
    "    ID = \"gene_1\",\n",
    "    chromosome = \"chr1\",\n",
    "    start = 1,\n",
    "    end = 100,\n",
    "    sense = \"+\",\n",
    "    multi = False\n",
    ")\n",
    "\n",
    "gene.add_exon(\n",
    "\n",
    "    key = \"exon_1\",\n",
    "    value = OD(\n",
    "        [   \n",
    "            (\"Start\", 1),\n",
    "            (\"End\", 10)\n",
    "        ]\n",
    "    )\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
