{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "import polars as pl \n",
    "import pandas as pd \n",
    "from data import NB_NT\n",
    "\n",
    "data = pl.read_csv(\"output/output.csv\", infer_schema_length = 10000)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_upstream(data_path = \"output/output.csv\", nb_codons : int = 3):\n",
    "\n",
    "    data = pl.read_csv(data_path, infer_schema_length = 10000).filter(pl.col(\"ribostartLocalisation\") == \"exon\")\n",
    "\n",
    "    threshold_f1 = ( nb_codons * 3 ) + 1\n",
    "    threshold_f2 = ( nb_codons * 3 ) + 2\n",
    "\n",
    "\n",
    "\n",
    "    sorted = data.filter(\n",
    "        \n",
    "        ((pl.col('upstream').apply(lambda x: len(x)) >= threshold_f1) & (pl.col('rel_frame') == 1)) |\n",
    "        ((pl.col('upstream').apply(lambda x: len(x)) >= threshold_f2) & (pl.col('rel_frame') == 2))\n",
    "\n",
    "    )\n",
    "\n",
    "    sorted = sorted.with_columns((pl.col('upstream') + pl.col('start_seq')).alias('sequence'))\n",
    "\n",
    "    codons = []\n",
    "\n",
    "\n",
    "    # Case F1 validated\n",
    "    for orf in sorted.filter(pl.col('rel_frame') == 1).iter_rows(named = True):\n",
    "        codons = []\n",
    "        for i in range(nb_codons):\n",
    "\n",
    "            ribostart = len(orf['sequence']) - 3\n",
    "            start = ribostart - 1 - 3 * i\n",
    "            end = ribostart + 2 - 3 * i\n",
    "            codons.append(orf['sequence'][start:end])\n",
    "    \n",
    "        \n",
    "\n",
    "    # Case F2 validated\n",
    "    for orf in sorted.filter(pl.col('rel_frame') == 2).iter_rows(named = True):\n",
    "        for i in range(nb_codons):\n",
    "\n",
    "            ribostart = len(orf['sequence']) - 3\n",
    "            start = ribostart - 2 - 3 * i\n",
    "            end = ribostart + 1 - 3 * i\n",
    "            codons.append(orf['sequence'][start:end])\n",
    "\n",
    "\n",
    "\n",
    "    return codons\n",
    "\n",
    "        \n",
    "def get_downstream(data_path = \"output/output.csv\", nb_codons : int = 3):\n",
    "\n",
    "    data = pl.read_csv(data_path, infer_schema_length = 10000).filter(pl.col(\"ribostartLocalisation\") == \"exon\")\n",
    "\n",
    "    threshold_f1 = ( nb_codons * 3 )  + 2\n",
    "    threshold_f2 = ( nb_codons * 3 ) + 1\n",
    "    \n",
    "\n",
    "    sorted = data.filter(\n",
    "        \n",
    "        ((pl.col('downstream').apply(lambda x: len(x)) >= threshold_f1) & (pl.col('rel_frame') == 1)) |\n",
    "        ((pl.col('downstream').apply(lambda x: len(x)) >= threshold_f2) & (pl.col('rel_frame') == 2))\n",
    "\n",
    "    )\n",
    "\n",
    "    sorted = sorted.with_columns((pl.col('start_seq') + pl.col('downstream')).alias('sequence'))\n",
    "\n",
    "\n",
    "    # Case F1 validated\n",
    "    for orf in sorted.filter(pl.col('rel_frame') == 1).iter_rows(named = True):\n",
    "\n",
    "        codons = []\n",
    "        for i in range(nb_codons):\n",
    "\n",
    "            ribostart = 0\n",
    "            start = ribostart + 2 + 3 * i\n",
    "            end = ribostart + 5 + 3 * i\n",
    "            codons.append(orf['sequence'][start:end])\n",
    "        print(f\"{orf['ID']}: {codons}\")\n",
    "\n",
    "\n",
    "    # Case F2 validated\n",
    "    for orf in sorted.filter(pl.col('rel_frame') == 2).iter_rows(named = True):\n",
    "\n",
    "        codons = []\n",
    "\n",
    "        for i in range(nb_codons):\n",
    "\n",
    "            ribostart = 0\n",
    "            start = ribostart + 1 + 3 * i\n",
    "            end = ribostart + 4 + 3 * i\n",
    "            codons.append(orf['sequence'][start:end])\n",
    "\n",
    "\n",
    "\n",
    "    return None       \n",
    "\n",
    "get_downstream()      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "from data import ORFS, RIBO, GFF_POLARS, ORF_DF_COLUMNS\n",
    "from utils import check_double_overlap\n",
    "\n",
    "ORFS = (\n",
    "        ORFS\n",
    "        .filter(pl.col(\"Type\") == \"nc_ovp_same-CDS\")\n",
    "        .with_columns([\n",
    "            pl.col(\"ID\").apply(lambda value: value.split(\"_\")[3]).alias(\"Phase\"), # Extract absolute strand phase from the ID generated by ORFMine\n",
    "            pl.lit(\"NA\").alias(\"Ovp_gene\") # Create a column for the gene ID that will be filled by check_double_overlap()\n",
    "        ])\n",
    "    )\n",
    "\n",
    "ORFS = ORFS.select(ORF_DF_COLUMNS).apply(check_double_overlap) # Use .apply() method to leverage parallelization\n",
    "ORFS.columns = ORF_DF_COLUMNS\n",
    "\n",
    "\n",
    "joined = ORFS.join(RIBO, left_on = \"ID\", right_on = \"Seq_ID\", how = \"inner\")\n",
    "\n",
    "test = joined.filter(\n",
    "    pl.col(\"MSMS\") == 1\n",
    ")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chrXI_-_135687-135866_1_nc_ovp_same-CDS']\n",
      "Two_or_more_genes\n",
      "\r"
     ]
    }
   ],
   "source": [
    "for overlapped_feature_name , data in test: # This loop returns the name by which data is groupped, and the data itself as a polars dataframe\n",
    "\n",
    "        # If overlapped feature is not a gene ( = transposable_element for example ) it's not stored\n",
    "        if \"gene\" not in GFF_POLARS.filter(pl.col(\"ID\") == overlapped_feature_name)[\"Type\"].unique().to_list(): \n",
    "\n",
    "            print(data[\"ID\"].to_list())\n",
    "            print(overlapped_feature_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
