{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Input genomic fasta file: Scer_SGD.fna\n",
      " Input gff file: Scer_SGD.gff\n",
      "Building structured data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 351379/351379 [00:01<00:00, 279792.74it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding Supplemental Attribute table...\n",
      "Finding unique attribute keys...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 351377/351377 [00:01<00:00, 323240.33it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making attribute table...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 351377/351377 [00:03<00:00, 114835.99it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "date Tue Jan 13 13:06:13 2015\n",
      " Created by Saccharomyces Genome Database (http://www.yeastgenome.org/)\n",
      " Weekly updates of this file are available for download from:\n",
      " http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff\n",
      "Building structured data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 23076/23076 [00:00<00:00, 281459.58it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding Supplemental Attribute table...\n",
      "Finding unique attribute keys...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 23058/23058 [00:00<00:00, 278089.40it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making attribute table...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 23058/23058 [00:00<00:00, 93561.85it/s]\n"
     ]
    }
   ],
   "source": [
    "import gff3_parser\n",
    "import polars as pl\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "orfs = pl.from_pandas(gff3_parser.parse_gff3(\"input/mapping_orf_Scer_SGD_noMT.gff\", parse_attributes=True))\n",
    "\n",
    "GFF = gff3_parser.parse_gff3(\"input/Scer.gff\", parse_attributes=True)\n",
    "\n",
    "FASTA = list(SeqIO.parse(\"input/Scer.fna\", \"fasta\"))\n",
    "\n",
    "\n",
    "COLUMNS = [[\"seq_id\", \"start\", \"end\", \"strand\", \"phase\", \"attributes\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import OrderedDict as OD\n",
    "import polars as pl\n",
    "from Bio.SeqIO import parse\n",
    "import re\n",
    "\n",
    "GFF_POLARS = pl.from_pandas(GFF)\n",
    "\n",
    "ORF_DF_COLUMNS = ['Seqid',\n",
    " 'Source',\n",
    " 'Type',\n",
    " 'Start',\n",
    " 'End',\n",
    " 'Score',\n",
    " 'Strand',\n",
    " 'Phase',\n",
    " 'Status',\n",
    " 'color',\n",
    " 'Parent',\n",
    " 'ID',\n",
    " 'Ovp_with',\n",
    " 'Ovp_gene']\n",
    "\n",
    "\n",
    "class GenomicFeature:\n",
    "\n",
    "    def __init__(self, ID : str, start : int, end : int):\n",
    "        self.ID = ID\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.counter = 0\n",
    "\n",
    "\n",
    "class Gene(GenomicFeature):\n",
    "\n",
    "    counter = 0\n",
    "\n",
    "    def __init__(self, ID, start, end, chromosome : str,  multi : bool, sense : str, frame : int):\n",
    "        super().__init__(ID, start, end)\n",
    "        self.chromosome = chromosome\n",
    "        self.frame = frame\n",
    "        self.sense = sense\n",
    "        self.multi = multi\n",
    "        self.aORFs = OD()\n",
    "        self._exons = OD()\n",
    "        Gene.counter += 1\n",
    "        \n",
    "\n",
    "    @property\n",
    "    def exons(self):\n",
    "        return self._exons\n",
    "    \n",
    "    @property\n",
    "    def exons_list(self):\n",
    "        return list(self._exons.values())\n",
    "\n",
    "    @exons.setter\n",
    "    def exons(self, value):\n",
    "    \n",
    "        self._exons = value\n",
    "\n",
    "    def add_exon(self, key, value):\n",
    "        \n",
    "        if key in self._exons:\n",
    "            raise KeyError(f'exon {key} already exists')\n",
    "\n",
    "        elif type(value) != Exon:\n",
    "\n",
    "            raise TypeError('exon must be an Exon')\n",
    "        \n",
    "        else:\n",
    "            self._exons[key] = value\n",
    "\n",
    "            \n",
    "    def add_orf(self, key, value):\n",
    "\n",
    "        if type(value) != Orf:\n",
    "            raise TypeError('value must be Orf')\n",
    "\n",
    "        elif key in self.aORFs:\n",
    "            raise KeyError(f'orf {key} already exists')\n",
    "        else:\n",
    "            self.aORFs[key] = value\n",
    "\n",
    "    \n",
    "    def sort(self):\n",
    "\n",
    "        if self.sense == \"+\":\n",
    "\n",
    "            if int(self.start) > int(self.end) : self.start, self.end = self.end, self.start\n",
    "            self._exons = OD(sorted(self._exons.items(), key=lambda x: x[1].start))\n",
    "\n",
    "        elif self.sense == \"-\":\n",
    "\n",
    "            if int(self.end) > int(self.start) : self.start, self.end = self.end, self.start\n",
    "            self._exons = OD(sorted(self._exons.items(), key=lambda x: x[1].start, reverse=True))\n",
    "\n",
    "\n",
    "class Exon(GenomicFeature):\n",
    "\n",
    "    def __init__(self, ID, start, end, gene : Gene, abs_frame : int):\n",
    "        super().__init__(ID, start, end )\n",
    "        self.abs_frame = abs_frame\n",
    "        self.gene = gene\n",
    " \n",
    "class Orf(GenomicFeature):\n",
    "\n",
    "    def __init__(self, ID, start, end, gene, ribospike : int):\n",
    "        super().__init__(ID, start, end)\n",
    "        self.gene = gene\n",
    "        self.ribospike = ribospike\n",
    "        self._ribostart = int(self.ribospike + self.start)\n",
    "        self.ribostartLocalisation = None\n",
    "        \n",
    "    def locRiboStart(self):\n",
    "\n",
    "        if self.gene.sense == \"+\":\n",
    "            for exon in self.gene.exons_list:\n",
    "                if exon.start <= self._ribostart <= exon.end:\n",
    "                    self.ribostartLocalisation = \"exon\"\n",
    "                    return\n",
    "                \n",
    "            if self.gene.start <= self._ribostart < self.gene.exons_list[0].start:\n",
    "                self.ribostartLocalisation = \"5UTR\"\n",
    "                return\n",
    "            \n",
    "            elif self.gene.exons_list[-1].end < self._ribostart <= self.gene.end:\n",
    "                self.ribostartLocalisation = \"3UTR\"\n",
    "                return\n",
    "\n",
    "            elif self._ribostart < self.gene.start:\n",
    "                self.ribostartLocalisation = \"upstream\"\n",
    "                return\n",
    "            \n",
    "            elif self._ribostart > self.gene.end:\n",
    "                self.ribostartLocalisation = \"downstream\"\n",
    "                return\n",
    "            \n",
    "            else:\n",
    "                self.ribostartLocalisation = \"intron\"\n",
    "                return\n",
    "            \n",
    "        elif self.gene.sense == \"-\":\n",
    "            for exon in self.gene.exons_list:\n",
    "                if exon.end <= self._ribostart <= exon.start:\n",
    "                    self.ribostartLocalisation = \"exon\"\n",
    "                    return\n",
    "                \n",
    "            if self.gene.exons_list[0].start < self._ribostart <= self.gene.start:\n",
    "                self.ribostartLocalisation = \"5UTR\"\n",
    "                return\n",
    "            \n",
    "            elif self.gene.end <= self._ribostart < self.gene.exons_list[-1].end:\n",
    "                self.ribostartLocalisation = \"3UTR\"\n",
    "                return\n",
    "\n",
    "            elif self._ribostart > self.gene.start:\n",
    "                self.ribostartLocalisation = \"upstream\"\n",
    "                return\n",
    "            \n",
    "            elif self._ribostart < self.gene.end:\n",
    "                self.ribostartLocalisation = \"downstream\"\n",
    "                return\n",
    "            \n",
    "            else:\n",
    "                self.ribostartLocalisation = \"intron\"\n",
    "                return\n",
    "            \n",
    "    \n",
    "\n",
    "    \n",
    "class GeneStructureError(Exception):\n",
    "    def __init__(self, message):\n",
    "        self.message = message\n",
    "        super().__init__(self.message)\n",
    "\n",
    "\n",
    "def get_good_column_names(columns):\n",
    "\n",
    "    import re\n",
    "    from itertools import chain\n",
    "\n",
    "    parent_col = [col for col in columns if re.match(r\"[Pp]arent\", col)]\n",
    "    name_col = [col for col in columns if re.match(r\"[Nn]ame\", col)]\n",
    "\n",
    "    if len(parent_col) == 1 and len(name_col) == 1:\n",
    "\n",
    "        return str(parent_col[0]), str(name_col[0])\n",
    "    \n",
    "    else:\n",
    "\n",
    "        raise AttributeError(\"Problem with GFF columns : Parent or Name not found. See get_good_column_names()\")\n",
    "        \n",
    "def return_gene_infos(gene_infos) -> dict:\n",
    "\n",
    "    chromosome = gene_infos[\"Seqid\"].unique().to_list()[0]\n",
    "    multi = gene_infos[\"Type\"].to_list().count(\"CDS\") > 1\n",
    "\n",
    "    gene_infos = gene_infos.filter(pl.col(\"Type\") == \"gene\")\n",
    "\n",
    "    start = gene_infos[\"Start\"].to_list()[0]\n",
    "    end = gene_infos[\"End\"].to_list()[0]\n",
    "    sense = gene_infos[\"Strand\"].to_list()[0]\n",
    "    \n",
    "    return {\n",
    "\n",
    "        \"chromosome\" : chromosome,\n",
    "        \"start\" : int(start),\n",
    "        \"end\" : int(end),\n",
    "        \"sense\" : sense,\n",
    "        \"multi\" : multi\n",
    "    } \n",
    "\n",
    "def init_gene_object(gene_id, gff_dataframe):\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    On veut pour la gene_id donnée initiliaser un objet Gene avec ses attributs\n",
    "    tirés du gff_dataframe.\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    pattern = fr\".*{gene_id}.*\"\n",
    "\n",
    "    parent, name = get_good_column_names(gff_dataframe.columns)\n",
    "\n",
    "    # Polars does not support regex filtering : pandas is used instead\n",
    "    gene_rows = pl.from_pandas(gff_dataframe[\n",
    "        gff_dataframe[parent].str.contains(pattern, regex=True, na=False) |\n",
    "        gff_dataframe[name].str.contains(pattern, regex=True, na=False)\n",
    "    ]) # Get all rows related to the gene being initialized\n",
    "\n",
    "\n",
    "    \n",
    "    gene_infos = return_gene_infos(gene_rows)\n",
    "\n",
    "\n",
    "    gene = Gene(\n",
    "\n",
    "        ID = gene_id,\n",
    "        chromosome = gene_infos[\"chromosome\"],\n",
    "        start = gene_infos[\"start\"],\n",
    "        end = gene_infos[\"end\"],\n",
    "        sense = gene_infos[\"sense\"],\n",
    "        multi = gene_infos[\"multi\"],\n",
    "        frame = gene_infos[\"start\"] % 3\n",
    "        \n",
    "    )\n",
    "    \n",
    "\n",
    "    cds_counter = 1\n",
    "    for exon in gene_rows.filter(pl.col(\"Type\") == \"CDS\").iter_rows(named = True):\n",
    "\n",
    "        key = f'{exon[\"ID\"]}-{cds_counter}' if exon[\"ID\"] else f'{exon[\"Name\"]}-{cds_counter}'\n",
    "        gene.add_exon(\n",
    "\n",
    "            key = key,\n",
    "            value = Exon(\n",
    "                        ID = key,\n",
    "                        gene = gene,\n",
    "                        start = int(exon[\"Start\"]),\n",
    "                        end = int(exon[\"End\"]),\n",
    "                        abs_frame = int(exon[\"Start\"]) % 3)\n",
    "        )\n",
    "\n",
    "        cds_counter = cds_counter + 1\n",
    "\n",
    "    gene.sort()\n",
    "\n",
    "    return gene\n",
    "\n",
    "def check_double_overlap(row : tuple):\n",
    "\n",
    "    orf = dict(zip(ORF_DF_COLUMNS, row))\n",
    "    \n",
    "    overlaps = [match for item in orf[\"Ovp_with\"].split(\"|\") for match in re.findall(r\"\\b([\\w-]+)_mRNA\\b\", item)]\n",
    "\n",
    "    if len(overlaps) == 0:\n",
    "\n",
    "        overlaps = [match for item in orf[\"Ovp_with\"].split(\"|\") for match in re.findall(r\"\\b([\\w-]+)_CDS\\b\", item)]\n",
    "    \n",
    "    buffer = []\n",
    "\n",
    "    if len(overlaps) != 1:\n",
    "\n",
    "        for overlap in overlaps:\n",
    "\n",
    "            \n",
    "            if GFF_POLARS.filter(\n",
    "                \n",
    "                (pl.col(\"ID\") == overlap)\n",
    "                )[\"Strand\"].unique().to_list()[0] == orf[\"Strand\"]:\n",
    "\n",
    "                buffer.append(overlap)\n",
    "                \n",
    "\n",
    "        if len(buffer) == 1: # Several genes are overlapped by the ORF, but only one is on the same strand\n",
    "\n",
    "            orf[\"Ovp_gene\"] = buffer[0]\n",
    "            return tuple(orf.values())\n",
    "        \n",
    "    \n",
    "        elif len(buffer) == 0: # No gene found on the same strand as the ORF\n",
    "\n",
    "            orf[\"Ovp_gene\"] = \"NA\"\n",
    "            return tuple(orf.values())\n",
    "        \n",
    "        else: # Several genes found on the same strand as the ORF\n",
    "\n",
    "            orf[\"Ovp_gene\"] = \"Two_or_more_genes\"\n",
    "            return tuple(orf.values())\n",
    "\n",
    "\n",
    "    else: # If there is only one gene found in the overlapping information given by ORFMine ID\n",
    "\n",
    "        orf[\"Ovp_gene\"] = overlaps[0]\n",
    "        return tuple(orf.values())\n",
    "\n",
    "def multifasta_to_dict(path, genome = False):\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    Reads a FASTA file and returns a dictionary where each record in the file\n",
    "    is a key-value pair with the record identifier as the key and the record sequence (in uppercase) as the value.\n",
    "\n",
    "    Args:\n",
    "    path (str): The path to the FASTA file to be read.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where the keys are the record identifiers and the values\n",
    "          are the corresponding record sequences as Seq objects in uppercase letters.\n",
    "    \"\"\"\n",
    "    \n",
    "    records = parse(path, \"fasta\")\n",
    "\n",
    "    if genome:\n",
    "        dico = {}\n",
    "        for record in records:\n",
    "            dico[record.id] = {}\n",
    "            dico[record.id][\"seq\"] = Seq(str(record.seq).upper())\n",
    "            dico[record.id][\"len\"] = len(record.seq)\n",
    "            \n",
    "        return dico\n",
    "    \n",
    "    else:\n",
    "\n",
    "        return {record.id: Seq(str(record.seq).upper()) for record in records}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern = r'\\b([\\w-]+)_CDS\\b'\n",
    "\n",
    "same_CDS_dframe = (\n",
    "    orfs\n",
    "    .filter(pl.col(\"Type\") == \"nc_ovp_same-CDS\")\n",
    "    .with_columns([\n",
    "        pl.col(\"ID\").apply(lambda value: value.split(\"_\")[3]).alias(\"Phase\"), # Extract absolute strand phase from the ID generated by ORFMine\n",
    "        pl.lit(\"NA\").alias(\"Ovp_gene\") # Create a column for the gene ID that will be filled by check_double_overlap()\n",
    "    ])\n",
    ")\n",
    "\n",
    "\n",
    "same_CDS_dframe = same_CDS_dframe.select(ORF_DF_COLUMNS).apply(check_double_overlap) # Use .apply() method to leverage parallelization\n",
    "\n",
    "same_CDS_dframe.columns = ORF_DF_COLUMNS\n",
    "\n",
    "grouped = same_CDS_dframe.groupby(\"Ovp_gene\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "intron\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "intron\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "intron\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "intron\n",
      "3UTR\n",
      "3UTR\n",
      "downstream\n",
      "3UTR\n",
      "3UTR\n",
      "intron\n",
      "3UTR\n",
      "3UTR\n",
      "intron\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "5UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "3UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "3UTR\n",
      "3UTR\n",
      "intron\n",
      "3UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "intron\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "downstream\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "exon\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "5UTR\n",
      "upstream\n",
      "exon\n",
      "upstream\n",
      "exon\n",
      "exon\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-207-79463b9216b4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m\"gene\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mGFF_POLARS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ID\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0moverlapped_feature_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Type\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# If overlapped feature is not a gene ( = transposable for example ) it's not stored\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m         \u001b[0mgene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_gene_object\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moverlapped_feature_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGFF\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miter_rows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnamed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-205-ae384314e38b>\u001b[0m in \u001b[0;36minit_gene_object\u001b[0;34m(gene_id, gff_dataframe)\u001b[0m\n\u001b[1;32m    227\u001b[0m     \u001b[0;31m# Polars does not support regex filtering : pandas is used instead\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m     gene_rows = pl.from_pandas(gff_dataframe[\n\u001b[0;32m--> 229\u001b[0;31m         \u001b[0mgff_dataframe\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mregex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    230\u001b[0m         \u001b[0mgff_dataframe\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mregex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m     ]) # Get all rows related to the gene being initialized\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/strings/accessor.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    127\u001b[0m                 )\n\u001b[1;32m    128\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 129\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    130\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    131\u001b[0m         \u001b[0mwrapper\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc_name\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/strings/accessor.py\u001b[0m in \u001b[0;36mcontains\u001b[0;34m(self, pat, case, flags, na, regex)\u001b[0m\n\u001b[1;32m   1258\u001b[0m             )\n\u001b[1;32m   1259\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1260\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_str_contains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcase\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mregex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1261\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_wrap_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturns_string\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/strings/object_array.py\u001b[0m in \u001b[0;36m_str_contains\u001b[0;34m(self, pat, case, flags, na, regex)\u001b[0m\n\u001b[1;32m    129\u001b[0m                 \u001b[0mupper_pat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m                 \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mupper_pat\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 131\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_str_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"bool\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    132\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    133\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_str_startswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/strings/object_array.py\u001b[0m in \u001b[0;36m_str_map\u001b[0;34m(self, f, na_value, dtype, convert)\u001b[0m\n\u001b[1;32m     69\u001b[0m         \u001b[0mmap_convert\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconvert\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 71\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer_mask\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muint8\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmap_convert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     72\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mTypeError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     73\u001b[0m             \u001b[0;31m# Reraise the exception if callable `f` got wrong number of args.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer_mask\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/strings/object_array.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    122\u001b[0m             \u001b[0mpat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mre\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 124\u001b[0;31m             \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msearch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    125\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    126\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcase\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "gene_list = list()\n",
    "FASTA_DICT = multifasta_to_dict(\"input/Scer.fna\", genome = True)\n",
    "\n",
    "for overlapped_feature_name , data in grouped: # This loop returns the name by which data is groupped, and the data itself as a polars dataframe\n",
    "\n",
    "    if \"gene\" in GFF_POLARS.filter(pl.col(\"ID\") == overlapped_feature_name)[\"Type\"].unique().to_list(): # If overlapped feature is not a gene ( = transposable for example ) it's not stored\n",
    "\n",
    "        gene = init_gene_object(overlapped_feature_name, GFF)\n",
    "\n",
    "        for row in data.iter_rows(named = True):\n",
    "\n",
    "            ribospike = 0 # For now, don't know how data will look like\n",
    "\n",
    "            orf = Orf(\n",
    "                ID = row[\"ID\"],\n",
    "                start = int(row[\"Start\"]),\n",
    "                end = int(row[\"End\"]),\n",
    "                gene = gene,\n",
    "                ribospike = ribospike\n",
    "                )\n",
    "\n",
    "            orf.locRiboStart()\n",
    "\n",
    "            gene.add_orf(key = orf.ID, \n",
    "                         value = orf)\n",
    "\n",
    "\n",
    "\n",
    "        gene_list.append(gene)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'> <class 'int'>\n",
      "<class 'int'> <class 'int'>\n",
      "<class 'int'> <class 'int'>\n"
     ]
    }
   ],
   "source": [
    "\n",
    "                      "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
