import polars as pl
import utils
from classes import Orf
from data import ORFS, GFF, GFF_POLARS, ORF_DF_COLUMNS, FASTA_DICT, RIBO, NB_NT
from halo import Halo



@Halo(text='Parsing input files ...\n', spinner='dots')
def parse_input():

    same_CDS_dframe = (
        ORFS
        .filter(pl.col("Type") == "nc_ovp_same-CDS")
        .with_columns([
            pl.col("ID").apply(lambda value: value.split("_")[3]).alias("Phase"), # Extract absolute strand phase from the ID generated by ORFMine
            pl.lit("NA").alias("Ovp_gene") # Create a column for the gene ID that will be filled by check_double_overlap()
        ])
    )


    same_CDS_dframe = same_CDS_dframe.select(ORF_DF_COLUMNS).apply(utils.check_double_overlap) # Use .apply() method to leverage parallelization

    same_CDS_dframe.columns = ORF_DF_COLUMNS

    joined = same_CDS_dframe.join(RIBO, left_on = "ID", right_on = "Seq_ID", how = "inner")


    return joined.groupby("Ovp_gene")


@Halo(text='Generating data ...\n', spinner='dots')
def extract_data(grouped):

    gene_list = list()

    for overlapped_feature_name , data in grouped: # This loop returns the name by which data is groupped, and the data itself as a polars dataframe

        # If overlapped feature is not a gene ( = transposable_element for example ) it's not stored
        if "gene" in GFF_POLARS.filter(pl.col("ID") == overlapped_feature_name)["Type"].unique().to_list(): 

            # Initialize gene object with the name of the overlapped feature
            gene = utils.init_gene_object(overlapped_feature_name, GFF)

            # Iterate over every ORF that overlaps the gene
            for row in data.iter_rows(named = True):

                orf = Orf(
                    ID = row["ID"],
                    start = int(row["Start"]),
                    end = int(row["End"]),
                    gene = gene,
                    ribospike = int(row["first.codon.idx"]),
                    MSMS = row["MSMS"],
                    MS_sds = row["msms.ynb.sds"],
                    MS_cond = row["msms.ynb.sds.mg132"],
                    age_rel = row["Relative.age"],
                    start_seq = row["first.codon"],
                    )

                orf.locRiboStart()

                gene.add_orf(key = orf.ID, 
                            value = orf)

            gene.get_adjacent_nucleotides(FASTA_DICT[gene.chromosome])
            gene_list.append(gene)
            
    print("\nNumber of ORFs processed : ")
    print(Orf.counter)

    return gene_list



